docs(tree): Undo v1 design doc (take 2) (#15930)
microsoft/FluidFramework: experimental/dds/tree2/docs/undo/v1-undo.md

## Repair Data

[Repair data](../repair-data/README.md) may be needed in the following cases:

-   Creating an undo (or redo) edit

> -   Applying a (peer or local) undo (or redo) edit on a `SharedTreeView`

Author: This line item is just referring to the standard undo redo stuff we already support (though with bugs). There must be something about the way it is phrased that's making it sound like something greater. Happy to chat about it if that's easier.

-   Applying the effect of an move (peer or local) that pulls content out from a previously deleted subtree and into the document tree
-   Applying the effect of rebasing local destructive changes
    (i.e., overwrites and deletions)
    that were predicated on a constraint that was violated as part of the rebase

Note that, while these are the only cases where a client needs to access the repair data for the subtree of interest,
it is necessary to keep the repair data be kept up to date until such a case arises.
This then forces us to consider cases where the repair data is not needed but is edited.
This can happen when a client edits a region of the document tree while that region is concurrently deleted by another client.

### The Stygian Forest

[Stygian](https://www.merriam-webster.com/dictionary/stygian),

> is used here to mean [Styx](https://en.wikipedia.org/wiki/Styx)-like in its ability to span the realms of in-document (living) content and deleted (dead) content,

Author: Ooh a haunting metaphor would be great here too!

> is used here to mean [Styx](https://en.wikipedia.org/wiki/Styx)-like in its ability to span the realms of in-document (living) content and deleted (dead) content,

Reviewer: Haha I would not be opposed to calling it the haunted forest. I was thinking Stygian requires a few more brain cells than I have.

### The Stygian Forest

[Stygian](https://www.merriam-webster.com/dictionary/stygian),
is used here to mean [Styx](https://en.wikipedia.org/wiki/Styx)-like in its ability to span the realms of in-document (living) content and deleted (dead) content,
and ferry data across it.

The responsibility of a `StygianForest` is to maintain,
in addition to the typical in-document data,
the repair data that may be needed to handle the cases listed above.

This coupling of repair data with the in-document data is motivated by the following points:

-   Aside from a pair of performance-motivated exceptions (covered below) the repair data is solely relevant to applying changes.
-   The needs of repair data storage, reading, and editing, are identical to that of in-document data.
-   Storing both in the same forest makes it easy, when it is appealing to do so,
    to efficiently remove/restore deleted content because it saves us from having to export/import it from and to the forest.

The alternative, coupling repair data with the changesets that birthed them, is undesirable for the following reasons:

> -   While, once rebased, a changeset is fixed, the repair data that this changeset initially produced may change over time.

Author: Yes, this does refer to the repair data changing due to possibly concurrent edits (I will expand on it for clarity).

So long as we're talking about a case where the move is targeting a portion of the deleted subtree, then yes, it makes the case where A is sequenced before B and the case where B is sequenced before A equivalent, meaning this approach makes those edits commutative. That's one of the benefits of it.

If we were talking about a case where the move is targeting the node that's concurrently being deleted, then it's a bit less clear what should happen, and we haven't made a final decision yet.

(i.e., storage, reading, and editing) for the purpose of managing repair data.
In effect, the `StygianForest` meets the more advanced requirements of our repair data system
(understanding the difference between normal document data vs. repair data, and supporting the more complex lifecycle of the latter)
by appropriately driving the forest it wraps.
This allows us to cleanly separate the concerns of locally storing, reading, and editing tree data on one side
from the the concerns of repair data on the other.

This approach to dealing with repair data requires that the `StygianForest` be able to maintain a correspondence between
how changesets refer to deleted/overwritten content
and how the inner forest refers to this same content.
Here is one approach that can be used:

-   Store the repair data under a detached field with a unique ID such as `styx-${counter}` where...
    -   `counter` is monotonically increased integer that is scope to the `StygianForest`.
-   Keep a `Map<RevisionId, { changeId: ChangesetLocalId, repairData: number }>` where..
    -   The map key is the revision associated with the changeset that lead to the creation of the repair data.

>     -   The `changeId` field is the `ChangesetLocalId` associate with the change that lead to the creation of a specific piece of repair data.

Author: I doesn't have to be, but in practice it is.

This includes paying attention to the `RevisionInfo.rollbackOf` field on changesets in order to correctly distinguish edits that should lead to the production of repair data from edits that should not.

### Garbage-Collecting Repair Data

The rules for the creation and consumption of repair data (see above) mean that,
while new content can be created,
no content is ever truly deleted.
This speaks to undo's role of preventing data loss, but leads to a problem of unbounded memory growth.

While we want to preserve the ability of a session to undo its past edits,
we also want to ensure that the associated memory needs can be constrained.
There are many ways this could be accomplished.
For now we adopt the following garbage-collection scheme:

The repair data associated with a given edit E must _not_ be garbage-collected if _either_ of the following criteria apply:

-   E was produced by the local session and E is within the undo window supported by the session.
-   E was still in the collaboration window
    (meaning its sequence number was >= to the min ref sequence number)

>     when the edit whose sequence number is the same as the current min ref sequence number.

Reviewer: I think this is missing something?

Author: Indeed.

This approach ensures that a client,
no matter how long its history of undoable/redoable edits,
does not burden its peers by requiring them to store unbounded amounts of repair data,
and does not force summaries to include that repair data.

The one drawback,
is that peers will not have the required repair data to apply the undo edit that a session might generate
if the undo occurs sufficiently late.
We address this problem by including in the undo commit being sent a copy of the relevant repair data,
if the edit being undone lies outside of the current collaboration window.
This acts as a "refresher" for any peers that may have garbage collected that repair data.
(See [Creating Repair Data On Change Application](#creating-repair-data-on-change-application) for details on how that refresher is handled by the peers.)

If this switch from not including the refresher with the undo to including it came any later,
meaning, if we did not include the refresher on undos created beyond the point where the undone edit falls out of the collaboration window,
then we would run the risk of peers receiving an undo for which they have already discarded the repair data:

> ![An undo edit is received by a peer that no longer has the repair data for it](./UndoAfterCollab.png)

Reviewer: I don't quite understand what the dark blue collab window includes. Could you add an example?

then we would run the risk of peers receiving an undo with incorrect repair data:

![An undo is sent with out of date repair data](./UndoDuringCollab.png)

In the scenario pictured above, the client sending the undo must not include a refresher.
If it did, and the green edit affected the portion of the document removed by the deletion,
then the repair data refresher would not reflect the impact of that edit.

It is safe for the undo sent in the scenario pictured above not to include the refresher.
This is due to the following invariant of the collaboration window:
If an edit is sent before a given collaboration window span S,
then it will be sequenced either before or within S, but no later.
Graphically speaking, this means that if an edit is sent for sequencing within the span of one rectangle,
then it will be sequenced at the latest within the span of rectangle that follows.
Here we can apply that rule to see that the undo edit,
so long as it is conceived within or before the light blue collaboration window span,
will be sequenced within or before the dark blue collaboration window span,

> during which peers will still be holding on to their copy of the repair data.

Reviewer: Leaving this comment here so I remember to go back to it once I understand what the dark blue means.

> While the repair data being send may be large,

Reviewer: ```suggestion
While the repair data being sent may be large,
```

and therefore costly to send over the wire,
this cost is only incurred when undo is actually used.

While this repair data is part of the undo commit,

> it is not a change properly speaking a change and therefore does not take part in rebasing operations.

Reviewer: ```suggestion
it is not properly speaking a change and therefore does not take part in rebasing operations.
```

(or at least all clients wanting such undo powers)
to retain the repair data associated with the peer edit to be undone.
While this applies more memory pressure on the clients,
it seems like a natural evolution of the system described in this document.

This new capability would however make possible a new family of scenarios:
two clients may concurrently attempt to undo the same edit.

If the latter of the two such undos carries with it a refresher,
then the receiving peers (as well as the issuing client) will end up
creating the repair data as the refresher instructs, despite the same tree content having already been revived.
This could be addressed by rendering refreshers ineffective when they are rebased over a revival of the associated subtree.
For example, the `StygianForest` could keep a record of the repair data entries
(but not the repair data itself)
that have been revived by the edits that lie within the current collaboration window.

#### Supporting Edits To Deleted Content

Instead of prohibiting edits to content that has been deleted (and whose deletion is known as opposed to concurrent),
we may want to allow such edits.

> This would give application more options for dealing with situations where a client has locally performed some amount work

Reviewer: ```suggestion
This would give applications more options for dealing with situations where a client has locally performed some amount of work
```

Unfortunately, #2 is needed because any edit made to the deleted subtree may entail moving content out of that subtree and into the document tree.
If clients that have garbage-collected the repair data for the subtree of interest have not been provided with a refresher before that,
then they will be unable to show an updated document tree.

This approach leads to the same "concurrent refresh" problem that [supporting undo of peer edits does](#supporting-undo-of-peer-edits),
presumably with a similar solution:
refreshers must be rendered ineffective when they are rebased over a concurrent revival or refresh of the associated subtree.

#### Accommodating The Destructuring of Nodes

There may be times when a client needs to replace a node with another, but keep some of the child nodes of the original.
More generally, a client may want to create a node and populate its fields with children from some other node.
This is easy to accomplish using the move editing API when the source and destination fields of children being transplanted allow for the absence of said children.
When either or both of those fields require the presence of a child
(as is the case for a value field)

> the it's either impossible to move the child out of the source field,

Reviewer: ```suggestion
then it's either impossible to move the child out of the source field,
```

This then forces us to consider cases where the repair data is not needed but is edited.
This can happen when a client edits a region of the document tree while that region is concurrently deleted by another client.

### Storing Repair Data in the Forest

The design presented here leverages the `Forest` owned by the `SharedTreeView` to house and maintain repair data.

> This made possible by the fact that all repair data comes in the shape of trees,

Reviewer: ```suggestion
This is made possible by the fact that all repair data comes in the shape of trees,
```

    making it more self-contained and therefore easier to implement and test.
    This is made possible by the fact that repair data is **not** required for the purpose of rebasing changesets.

### The Repair Data `TreeIndex`

One challenge associated with storing repair data in the forest,
is that it requires us to establish a system for keeping track of which removed subtrees exist in the repair data,
and where they reside in the forest.
Specifically, we have to maintain a correspondence between the identification scheme used by changesets, and the one used by the forest.

> Changesets identify removed trees using a `ChangeAtomId` while the forest uses paths.

Reviewer: What's the meaning behind this name?

Author: When we have a change that deletes 3 nodes in a row in a sequence field, we represent that with a single Delete mark of length 3, but that's just an optimisation for the fact that there are 3 deletions. We've used the term "change atom" to refer the the smallest unit of change that a changeset can be made of. In that "delete 3 nodes" scenario, we allocate 3 `ChangeAtomId`s for the operation, but because we allocate them as one block, we only have to write down the `ChangeAtomId` for the first deletion and the rest can be infered. Identifying the change in this way also lets us identify the affected nodes using the same ID.

In order to meet this need, we introduce the `TreeIndex` class,
which maintains a mapping between the two identification systems,

> and also serves as inventory of the repair data that exists in the forest.

Reviewer: ```suggestion
and also serves as an inventory of the repair data that exists in the forest.
```

We also introduce the concept of `RemovedTreeId`,
which is a value type (likely a number) that uniquely identifies a removed subtree within the forest on a given `SharedTreeView`.

`TreeIndex`'s interface needs to offer the following methods:

-   Retrieve the matching `RemovedTreeId` for a given `ChangeAtomId` if any.
-   Add an entry in its records for a given `ChangeAtomId` to `RemovedTreeId` mapping.
-   Remove an entry from its records for a given `ChangeAtomId`.

The first of these methods can be kept under a separate `ReadonlyTreeIndex` interface that `TreeIndex` extends.

> The generation of new `RemovedTreeId`s and their conversion to forest paths is left as implementation detail for the `SharedTreeView`.

Reviewer: ```suggestion
The generation of new `RemovedTreeId`s and their conversion to forest paths is left as an implementation detail of the `SharedTreeView`.
```

In the future, it is likely that the `Forest` will take on these responsibilities.
For now, the `SharedTreeView` can use a monotonically increasing counter to generate `RemovedTreeId`s,

> and can use `removed-${removedTreeId}` as way of generating unique paths.

Reviewer: ```suggestion
and can use `removed-${removedTreeId}` as a way of generating unique paths.
```

-   When a subtree is deleted
-   When the value on an node is overwritten

When a changeset is applied to the `SharedTreeView`,
for every change conveyed by the changeset that would erase document data,
the `SharedTreeView` must do the following:

-   Generate a new `RemovedTreeId`.
-   Translate that change into an equivalent delta that preserves the otherwise deleted subtree
    by moving that subtree to a part of the inner forest that lies outside the scope of the document.
-   Apply that delta to the forest.
-   Update the `TreeIndex` so it has an entry associating the removed subtree's `ChangeAtomId` and `RemovedTreeId`.

Another scenario where repair data is added to the `Forest`
arises when a peer sends a commit whose changes explicitly instructs the `SharedTreeView` to do so,
and provides a `ChangeAtomId` as well as the tree contents.
When that happens, `SharedTreeView` should consult its `TreeIndex` to check whether an entry already exists for the `ChangeAtomId`.
If it does, then it can safely keep the repair data that it has.
If it doesn't, then it should:

-   Generate a new `RemovedTreeId`.

> -   Generate a delta a delta that adds the content to the forest.

Reviewer: ```suggestion
-   Generate a delta that adds the content to the forest.
```

    for each piece of repair data associated with changesets that make up the transaction,
    leave the repair data in the `StygianForest` untouched but update its entry so that it is now associated with the composed changeset.

> Note that the `AnchorSet` for the view also needs to be patched so that any anchors pointing to the repair data is updated with the appropriate path.

Reviewer: ```suggestion
Note that the `AnchorSet` for the view also needs to be patched so that any anchors pointing to the repair data are updated with the appropriate path.
```

leading to unnecessary state churning and invalidation of downstream components.

Another approach that could be taken, would be to clone the `SharedTreeView` components
(such as the `Forest`, `TreeIndex`, and `AnchorSet`)
of the branch onto which the current `SharedTreeView` is being rebased,
and apply the rebased edits to that clone.
The drawback of that approach is that this not only may cause invalidation in parts of the documents that are not actually changed by the rebase,
but it also forces all consumers of these components to update/rebuild any references that might have had (such as anchors) to those components.

> The upshot is that `SharedTreeView` must tolerate the kind of changesets that results from such rebase operations.

Reviewer: ```suggestion
The upshot is that `SharedTreeView` must tolerate the kind of changesets that result from such rebase operations.
```
